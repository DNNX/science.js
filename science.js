(function(){science = {version: "0.0.2"}; // semver
science.functor = function(v) {
  return typeof v === "function" ? v : function() { return v; };
};
})()