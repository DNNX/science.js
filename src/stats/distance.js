science.stats.distance = {
  euclidean: function(a, b) {
    var n = a.length,
        i = -1,
        s = 0,
        x;
    while (++i < n) {
      x = a[i] - b[i];
      s += x * x;
    }
    return Math.sqrt(s);
  },
  manhattan: function(a, b) {
    var n = a.length,
        i = -1,
        s = 0;
    while (++i < n) s += Math.abs(a[i] - b[i]);
    return s;
  },
  minkowski: function(p) {
    return function(a, b) {
      var n = a.length,
          i = -1,
          s = 0;
      while (++i < n) s += Math.pow(Math.abs(a[i] - b[i]), p);
      return Math.pow(s, 1 / p);
    };
  },
  chebyshev: function(a, b) {
    var n = a.length,
        i = -1,
        max = 0,
        x;
    while (++i < n) {
      x = Math.abs(a[i] - b[i]);
      if (x > max) max = x;
    }
    return max;
  },
  hamming: function(a, b) {
    var n = a.length,
        i = -1,
        d = 0;
    while (++i < n) if (a[i] !== b[i]) d++;
    return d;
  },
  jaccard: function(a, b) {
    var n = a.length,
        i = -1,
        s = 0;
    while (++i < n) if (a[i] === b[i]) s++;
    return s / n;
  }
};
