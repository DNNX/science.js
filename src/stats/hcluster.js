science.stats.hcluster = function() {
  var distance = science.stats.distance.euclidean;

  function hcluster(vectors) {
    var n = vectors.length;
    var dMin = [];
    var cSize = [];

    var matrixObj = new figue.Matrix(N,N);
    var distMatrix = matrixObj.mtx ;
    var clusters = [];

    var c1, c2, c1Cluster, c2Cluster, i, j, p, root , newCentroid ;

    var i,
        j;

    // Initialize distance matrix and vector of closest clusters
    i = -1; while (++i < n) {
      dMin[i] = 0;
      j = -1; while (++j < n) {
        distMatrix[i][j] = i === j ? Infinity : distance(vectors[i] , vectors[j]);
        if (distMatrix[i][dMin[i]] > distMatrix[i][j]) dMin[i] = j ;
      }
    }

    // create leaves of the tree
    i = -1; while (++i < n) {
      clusters[i] = [];
      clusters[i][0] = new Node(labels[i], null, null, 0, vectors[i]);
      cSize[i] = 1;
    }

    // Main loop
    for (p = 0; p < N-1; p++) {
      // find the closest pair of clusters
      c1 = 0 ;
      for (i = 0 ; i < N ; i++) {
        if (distMatrix[i][dMin[i]] < distMatrix[c1][dMin[c1]]) c1 = i;
      }
      c2 = dMin[c1];

      // create node to store cluster info 
      c1Cluster = clusters[c1][0] ;
      c2Cluster = clusters[c2][0] ;

      newCentroid = calculateCentroid ( c1Cluster.size , c1Cluster.centroid , c2Cluster.size , c2Cluster.centroid ) ;
      newCluster = new Node (-1, c1Cluster, c2Cluster , distMatrix[c1][c2] , newCentroid) ;
      clusters[c1].splice(0,0, newCluster) ;
      cSize[c1] += cSize[c2] ;

      // overwrite row c1 with respect to the linkage type
      for (j = 0 ; j < N ; j++) {
              if (linkage == figue.SINGLE_LINKAGE) {
                      if (distMatrix[c1][j] > distMatrix[c2][j])
                              distMatrix[j][c1] = distMatrix[c1][j] = distMatrix[c2][j] ;
              } else if (linkage == figue.COMPLETE_LINKAGE) {
                      if (distMatrix[c1][j] < distMatrix[c2][j])
                              distMatrix[j][c1] = distMatrix[c1][j] = distMatrix[c2][j] ;
              } else if (linkage == figue.AVERAGE_LINKAGE) {
                      var avg = ( cSize[c1] * distMatrix[c1][j] + cSize[c2] * distMatrix[c2][j])  / (cSize[c1] + cSize[j]) 
                      distMatrix[j][c1] = distMatrix[c1][j] = avg ;
              }
      }
      distMatrix[c1][c1] = Infinity;

      // infinity Â­out old row c2 and column c2
      for (i = 0 ; i < N ; i++)
        distMatrix[i][c2] = distMatrix[c2][i] = Infinity;

      // update dmin and replace ones that previous pointed to c2 to point to c1
      for (j = 0; j < N ; j++) {
        if (dMin[j] == c2)
          dMin[j] = c1;
        if (distMatrix[c1][j] < distMatrix[c1][dMin[c1]]) 
          dMin[c1] = j;
      }

      // keep track of the last added cluster
      root = newCluster;
    }

    return root;
  }

  return hcluster;
};
