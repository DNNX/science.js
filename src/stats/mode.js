science.stats.mode = function(x) {
  if ((n = x.length) === 1) return x[0];
  x = x.slice().sort(science.ascending);
  var mode,
      n,
      i = -1,
      l = i,
      last,
      max = 0,
      tmp,
      v;
  while (++i < n) {
    if ((v = x[i]) !== last) {
      if ((tmp = i - l) > max) {
        max = tmp;
        mode = last;
      }
      last = v;
      l = i;
    }
  }
  return mode;
};
