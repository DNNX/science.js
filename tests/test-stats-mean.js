require("./../lib/env-js/envjs/node");
require("./../science");
require("./../science.stats");

var data = [1, 2, 3, 4, 5];

for (var i = 0; i <= data.length; i++) {
  var d = data.slice(0, i);
  console.log("mean [" + d + "]:");
  console.log("  ", science.stats.mean(d));
  console.log("");
}
