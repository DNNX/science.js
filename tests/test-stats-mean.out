mean []:
   NaN

mean [1]:
   1

mean [1,2]:
   1.5

mean [1,2,3]:
   2

mean [1,2,3,4]:
   2.5

mean [1,2,3,4,5]:
   3

